### PR TITLE
Implement default encoder and residual vq for autoencoder

### LINKED ISSUE #11: Make Residual Vector Quantization Block in AutoEncoder

### LINKED ISSUE #12: Default Auto Encoder with reconstruction loss

### PR DESCRIPTION
**Pull request by bwook00**
close #11 
close #12 

While review, you don't need to read full codes.
Because I got this code from below repo

- Encoder from: https://github.com/hrbigelow/ae-wavenet
- RVQ from: https://github.com/lucidrains/vector-quantize-pytorch

If something wrong while training, I will fix it later.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -27,8 +27,10 @@ dependencies = [
     "setuptools-scm>=8.2.0",
     "sounddevice>=0.5.1",
     "soundfile>=0.13.1",
+    "tensorflow>=2.19.0",
```
Comment (vkehfdl1) [inline pyproject.toml]: Why use tensorflow?!?!??!?!?!?!?!?!

File: songsuro/autoencoder/encoder/encoder.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
+import torch
+from torch import nn
+from .vconv import VirtualConv, output_offsets
+from .netmisc import xavier_init
```
Comment (vkehfdl1) [inline songsuro/autoencoder/encoder/encoder.py:4]: Recommend to not use relative path import
It can be occur some issues when users install it via pypi packages.

File: songsuro/autoencoder/loss.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import torch.nn as nn
+
+
+def reconstruction_loss(original, reconstructed):
+	recon_loss_fn = nn.L1Loss()
+	return recon_loss_fn(reconstructed, original)
```
Comment (vkehfdl1) [inline songsuro/autoencoder/loss.py:6]: I think we just use nn.L1Loss() directly??

File: songsuro/autoencoder/models.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import torch.nn as nn
+
+
+class Autoencoder(nn.Module):
+	def __init__(self, encoder, quantizer, decoder):
+		super(Autoencoder, self).__init__()
```
Comment (vkehfdl1) [inline songsuro/autoencoder/models.py:6]: Any difference between `super().__init__()`?

**Review by vkehfdl1 — CHANGES_REQUESTED** [review]

File: songsuro/autoencoder/models.py (hunk 4 above)
Comment (bwook00) [reply songsuro/autoencoder/models.py:6]: i'll fix at later issue

**Review by vkehfdl1 — APPROVED** [review]
LGTM

**Review by minsing-jin — APPROVED** [review]
LGTM

image
